---
**Autor**: Profs. Dino Magri e João Nogueira

**Contato**: `professor.dinomagri@gmail.com`, `joaonogueira@fisica.ufc.br`

**Licença deste notebook**: 
<br>
<img align="left" width="80" src="https://licensebuttons.net/l/by/3.0/88x31.png" />

<br>
<br>

[Clique aqui para saber mais sobre a licença CC BY v4.0](https://creativecommons.org/licenses/by/4.0/legalcode.pt)


---

# Parte 2 - Análise Exploratória de Dados

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('/content/drive/MyDrive/projeto-ia-codigos/projeto-ia-datasets/preco_casas_final.csv', sep=';')

In [ ]:
# Essa opção do pandas formata a visualização dos valores das colunas com o tipo 
# float (decimal) com duas casas decimais.
# pd.options.display.float_format = "{:.2f}".format

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
df.head()

zoneamento  tam_terreno forma_terreno  ...  mes_venda  ano_venda      preco
0         RL    785.03035           Reg  ...          2       2008   846510.0
1         RL    891.86880           Reg  ...          5       2007   661780.0
2         RL   1045.15875           IR1  ...          9       2008   907410.0
3         RL    887.22365           IR1  ...          2       2006   568400.0
4         RL   1324.79678           IR1  ...         12       2008  1015000.0

[5 rows x 23 columns]

### Variável Qualitativa e Variável Quantitativa

**Variável Qualitativa Nominal**

Quando não há ordenação entre valores!

In [6]:
df.zoneamento.unique()

array(['RL', 'RM', 'C (all)', 'FV', 'RH'], dtype=object)

**Variável Qualitativa Ordinal**

Quando há ordenação entre os valores.

In [ ]:
df.qualidade_cozinha.unique()

array(['Gd', 'TA', 'Ex', 'Fa', 'missing'], dtype=object)

In [ ]:
df.dtypes

zoneamento              object
tam_terreno            float64
forma_terreno           object
qualidade_geral          int64
condicao                 int64
ano_construcao           int64
qualidade_aquecedor     object
ar_condicionado         object
tam_primeiro_andar     float64
tam_segundo_andar      float64
tam_sala_estar         float64
qtde_quartos             int64
qualidade_cozinha       object
qtde_banheiros           int64
qtde_comodos             int64
qtde_lareiras            int64
qtde_carros_garagem      int64
tam_garagem            float64
tam_piscina            float64
mes_venda                int64
ano_venda                int64
preco                  float64
dtype: object

Tanto `qualidade_cozinha` quanto `zoneamento` são do tipo objeto. Podemos então utilizar o método `describe(include='object')` para obtermos estatísticas dessas variáveis: 

In [7]:
df[['zoneamento', 'qualidade_cozinha']].describe()

zoneamento qualidade_cozinha
count        1458              1458
unique          5                 4
top            RL                TA
freq         1149               735

Podemos também com o mesmo método obter informações de todas as variáveis qualitativas de uma só vez:

In [ ]:
df.describe(include='object')

zoneamento forma_terreno  ... ar_condicionado qualidade_cozinha
count        1459          1459  ...            1459              1459
unique          5             4  ...               2                 5
top            RL           Reg  ...               Y                TA
freq         1150           924  ...            1364               735

[4 rows x 5 columns]

**Variável Quantitativa Discreta**

Assumem valores dentro de um conjunto finito.

In [ ]:
df.qtde_quartos.unique()

array([3, 4, 1, 2, 0, 5, 6, 8])

**Variável Quantitativa Contínua**

Assumem valores em uma escala contínua.

In [ ]:
df.preco[:4]

0    846510.000000
1    734478.354431
2    907410.000000
3    568400.000000
Name: preco, dtype: float64

### Medidas de posição

O método describe fornece as estatísticas descritivas (```quantidade, média, desvio padrão, mínimo, máximo, percentis```) que apresentam a tendência central, a dispersão e a forma da distribuição do conjunto de dados.

In [ ]:
df.describe()

tam_terreno  qualidade_geral  ...    ano_venda         preco
count   1459.000000      1459.000000  ...  1459.000000  1.459000e+03
mean     976.860145         6.098012  ...  2007.816998  7.341219e+05
std      927.580526         1.382574  ...     1.327699  3.223512e+05
min      120.773900         1.000000  ...  2006.000000  1.416940e+05
25%      701.324747         5.000000  ...  2007.000000  5.275970e+05
50%      880.441731         6.000000  ...  2008.000000  6.617800e+05
75%     1077.674800         7.000000  ...  2009.000000  8.688400e+05
max    19996.906235        10.000000  ...  2010.000000  3.065300e+06

[8 rows x 17 columns]

Para as medidas Moda e Mediana, podemos aplicá-las em uma ou várias colunas especificas.

In [ ]:
df.qtde_quartos.mode()

0    3
dtype: int64

In [ ]:
df[['qtde_quartos', 'qtde_lareiras']].median()

qtde_quartos     3.0
qtde_lareiras    1.0
dtype: float64

### Medidas de dispersão

A finalidade é encontrar um valor que resuma a variabilidade de um conjunto de dados. Essas medidas indicam o quão espalhados os dados estão, ou como eles variam.

In [8]:
df_desc = df.describe()

#### Amplitude

Diferença entre o valor máximo e o valor mínimo.

In [9]:
df_desc.loc['max'] - df_desc.loc['min']

tam_terreno            1.987613e+04
qualidade_geral        9.000000e+00
condicao               8.000000e+00
ano_construcao         1.380000e+02
tam_primeiro_andar     4.048713e+02
tam_segundo_andar      1.918447e+02
tam_sala_estar         4.931291e+02
qtde_quartos           8.000000e+00
qtde_banheiros         3.000000e+01
qtde_comodos           1.200000e+01
qtde_lareiras          3.000000e+00
qtde_carros_garagem    4.000000e+00
tam_garagem            1.317365e+02
tam_piscina            6.856241e+01
mes_venda              1.100000e+01
ano_venda              4.000000e+00
preco                  2.923606e+06
dtype: float64

#### Intervalo Interquartil

É a diferença entre o terceiro e o primeiro quartil (Q3 - Q1)

In [ ]:
df_desc.loc['75%'] - df_desc.loc['25%']

tam_terreno               376.350053
qualidade_geral             2.000000
condicao                    1.000000
ano_construcao             46.000000
tam_primeiro_andar         47.334078
tam_segundo_andar          67.633384
tam_sala_estar             60.108241
qtde_quartos                1.000000
qtde_banheiros              1.000000
qtde_comodos                2.000000
qtde_lareiras               1.000000
qtde_carros_garagem         1.000000
tam_garagem                22.575429
tam_piscina                 0.000000
mes_venda                   3.000000
ano_venda                   2.000000
preco                  341243.000000
dtype: float64

#### Variância 

A variância é uma medida que expressa quanto os dados de um conjunto estão afastados de seu valor esperado. O cálculo é feito pela média dos quadrados dos desvios em relação a média aritmética.

Quanto maior for a variância, mais distantes da média estarão os valores, e quanto menor for a variância, mais próximos os valores estarão da média. As vezes é importante visualizar essa discrepancia seja mais perceptível e por isso utiliza-se a variância.

In [ ]:
df.var()

tam_terreno            8.604056e+05
qualidade_geral        1.911512e+00
condicao               1.238945e+00
ano_construcao         9.121057e+02
tam_primeiro_andar     1.290778e+03
tam_segundo_andar      1.640212e+03
tam_sala_estar         2.379705e+03
qtde_quartos           6.650683e-01
qtde_banheiros         8.564005e-01
qtde_comodos           2.639487e+00
qtde_lareiras          4.157770e-01
qtde_carros_garagem    5.578195e-01
tam_garagem            3.945322e+02
tam_piscina            1.394176e+01
mes_venda              7.309686e+00
ano_venda              1.762784e+00
preco                  1.039103e+11
dtype: float64

In [ ]:
df.mean()

tam_terreno               976.860145
qualidade_geral             6.098012
condicao                    5.575737
ano_construcao           1971.245374
tam_primeiro_andar        108.013460
tam_segundo_andar          32.174684
tam_sala_estar            140.731489
qtde_quartos                2.865661
qtde_banheiros              1.583276
qtde_comodos                6.516107
qtde_lareiras               0.612748
qtde_carros_garagem         1.766278
tam_garagem                43.927391
tam_piscina                 0.256486
mes_venda                   6.320082
ano_venda                2007.816998
preco                  734121.903680
dtype: float64

**Desvio Padrão**

O desvio padrão também é uma medida de dispersão e indica o quanto os dados estão adastados da média. Ao contrário da variância, em alguns cenários não é necessário que a discrpancia seja perceptível, pois esse valor pode ser considerado com um erro. E desta forma, se utilizassemos a variância para o erro, os valores mais discrepantes da média teriam um erro desproporcional e prejudicaria o experimento em questão.

Nesse cenário que podemos utilizar o desvio padrão.

O cálculo do desvio padrão é dado pela raiz da variância.

In [ ]:
df.std()

tam_terreno               927.580526
qualidade_geral             1.382574
condicao                    1.113079
ano_construcao             30.201088
tam_primeiro_andar         35.927398
tam_segundo_andar          40.499526
tam_sala_estar             48.782217
qtde_quartos                0.815517
qtde_banheiros              0.925419
qtde_comodos                1.624650
qtde_lareiras               0.644808
qtde_carros_garagem         0.746873
tam_garagem                19.862835
tam_piscina                 3.733866
mes_venda                   2.703643
ano_venda                   1.327699
preco                  322351.178121
dtype: float64

In [ ]:
df.preco.var()

103910282035.9178

In [ ]:
import numpy as np

In [ ]:
np.sqrt(df.preco.var())

322351.17812087765

**Coeficiente de variação**

Mede a variabilidade numa escala percentual independente da unidade de medida ou da ordem de grandeza da variável.

In [ ]:
cv = (df.preco.std() / df.preco.mean()) * 100
print(cv)

43.909761649267814


### Tabulação Cruzada (crosstab)

Podemos utilizar essa funcionalidade para cruzar os valores das variáveis.

In [ ]:
pd.crosstab(df.qtde_quartos, df.condicao)

condicao      1  2  3   4    5    6    7   8   9
qtde_quartos                                    
0             0  0  0   1    3    1    0   0   1
1             1  1  3   1   31    5    3   5   0
2             0  2  7  14  195   63   57  17   3
3             0  1  7  24  462  149  112  38  11
4             0  1  6  13  117   31   28   9   7
5             0  0  2   2    8    2    4   3   0
6             0  0  0   2    4    1    0   0   0
8             0  0  0   0    0    0    1   0   0

Analisando a tabela acima vemos que o **maior** número de imoveis na condição 5 são os imóveis com 3 quartos (462).

## Exercícios para explorar os dados

Para saber a quantidade de imóveis pela quantidade de banheiros, podemos utilizar o comando abaixo que realizar a contagem.

In [ ]:
pd.value_counts(df.qtde_banheiros)

2     767
1     650
3      32
0       9
30      1
Name: qtde_banheiros, dtype: int64

**<span style="color:blue">[Responda]</span>**  Quantos imóveis tem 3 quartos?

In [11]:
pd.value_counts(df.qtde_banheiros == 3)

False    1426
True       32
Name: qtde_banheiros, dtype: int64

**<span style="color:blue">[Responda]</span>**  Quantos imóveis têm 2 banheiros?

In [13]:
pd.value_counts(df.qtde_banheiros == 2)

True     767
False    691
Name: qtde_banheiros, dtype: int64

**<span style="color:blue">[Responda]</span>**  Quantos imóveis tem 3 banheiros e 3 quartos?

In [ ]:
pd.value_counts(df.qtde_banheiros == 3)

Podemos criar novas colunas que podem facilitar o nosso entendimento dos dados!

Por exemplo, quantas casas são pequenas, médias e grandes? 
- Como podemos adicionar essa informações dentro do nosso conjunto de dados? 
- Qual coluna disponível podemos utilizar?

**<span style="color:blue">[Responda]</span>** Crie uma função que identifica o tamanho do terreno com as seguintes condições:

    - pequeno: se menor ou igual 500
    - media: se maior que 500 e menor que 1000
    - grande: se maior ou igual a 1000

In [ ]:
def categorizar(tam):
  # adicione seu codigo aqui

In [ ]:
# Testando
print(categorizar(50))
print(categorizar(501))
print(categorizar(5000))

Crie uma nova coluna no DataFrame chamado 'categoria_terreno' e aplique a função criada anteriormente para todas as linhas do DataFrame. Lembre-se do método .apply que existe em uma Serie.

**<span style="color:blue">[Responda]</span>** Imprima as 5 primeiras linhas do DataFrame `df` apenas para as colunas `tam_terreno` e `categoria_terreno`.

**<span style="color:blue">[Responda]</span>**  Quantos imóveis tem o terreno pequeno?

**<span style="color:blue">[Responda]</span>**  Quantos imóveis tem o terreno médio?

**<span style="color:blue">[Responda]</span>** Quantos imóveis tem o terreno grande?